base/lock.jl

```diff
@@ -201,6 +201,14 @@ function trylock(f, l::AbstractLock)
     return false
 end
 
+"""
+    @lock l expr
+
+Macro version for `lock(f, l::AbstractLock)`. It wraps an arbitrary `expr` with a pair of
+`lock(l)` and `unlock(l)` functions. It is often more performant than function form
+`lock(f, l)`, because it don't capture variables into a lambda.
+See [`lock`](@ref).
```

Comment:
Perhaps something like this? Just a suggestion, but since the expansion is pretty short I think it is nice to show that no "magic" is happening under the hood.
````suggestion
Lock the lock `l`, evaluate `expr` inside [`try`](@ref), and unlock `l`. Expands to:
```julia
lock(l)
try
    expr
finally
    unlock(l)
end
```
This is similar to using [`lock`](@ref) with a `do` block:
```julia
lock(l) do
    expr
end
```
but avoids creating a closure.
````

Reply:
Lock the lock `l`, evaluate `expr` inside [`try`](@ref), and unlock `l`.
The same docstring is just above this one, for a `lock` function. So maybe we should just reference to it and add only `delta` docstring?

Reply:
Should we provide examples of expanded macro? I didn't see any of that in docs before with any other macro.

Reply:
I just thought it was short enough that it could be included, and including it is by far the easiest way to explain what is happening IMO but the full `lock`-`do` example might not be needed. Perhaps like this?

````suggestion
Lock the lock `l`, evaluate `expr` inside [`try`](@ref), and unlock `l`. Expands to:
```julia
lock(l)
try
    expr
finally
    unlock(l)
end
```
This is similar to using [`lock`](@ref) with a `do` block but avoids creating a closure.
````

Reply:
How about just sending user to `lock` docs? From next commit:
```julia
"""
Macro version for `lock(f, l::AbstractLock)`, but with `expr` instead of `f` function.
This is similar to using [`lock`](@ref) with a `do` block, but avoids creating a closure
and thus can improve the performance.
"""
```

Reply:
What's wrong with including the expansion in the docstring? Then it becomes self-explanatory. I don't understand what "macro version of lock" means.

Reply:
Just never seen code expansion in other macro docstrings, so I thought that `macroexpand` should be used for this.
I have included the expansion now. Should I also change "macro version of lock"? For example, to this: "Equivalent to lock but ..."